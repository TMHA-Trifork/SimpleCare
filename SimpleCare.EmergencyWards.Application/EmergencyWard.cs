namespace SimpleCare.EmergencyWards.Application;

public class EmergencyWard;
