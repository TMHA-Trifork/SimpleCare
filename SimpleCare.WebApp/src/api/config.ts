export const API_CONFIG = {
  baseUrl: 'http://localhost:5034'
}

export function getApiUrl(path: string): string {
  return `${API_CONFIG.baseUrl}${path}`
}
