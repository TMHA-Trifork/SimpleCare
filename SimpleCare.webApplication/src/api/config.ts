export const API_CONFIG = {
  baseUrl: 'https://localhost:7023'
}

export function getApiUrl(path: string): string {
  return `${API_CONFIG.baseUrl}${path}`;
}
