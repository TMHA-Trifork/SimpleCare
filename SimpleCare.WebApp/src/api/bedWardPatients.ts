import { components } from '../types/api-types'
import { getApiUrl } from './config'

export type BedWardPatientListItem = components['schemas']['BedWardPatientListItem']
export type BedWardPatient = components['schemas']['BedWardPatient']
export type PatientAdmission = components['schemas']['PatientAdmission']

export async function fetchBedWardPatients(): Promise<BedWardPatientListItem[]> {
  const response = await fetch(getApiUrl('/api/bed-ward/patients'))
  if (!response.ok) {
    throw new Error('Network response was not ok')
  }
  return response.json()
}

export async function fetchBedWardPatient(patientId: string): Promise<BedWardPatient> {
  const response = await fetch(getApiUrl(`/api/bed-ward/patients/${patientId}`))
  if (!response.ok) {
    throw new Error('Network response was not ok')
  }
  return response.json()
}

export async function admitPatient(admission: PatientAdmission): Promise<void> {
  const response = await fetch(getApiUrl('/api/bed-ward/patients/admit'), {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(admission)
  })
  if (!response.ok) {
    throw new Error('Network response was not ok')
  }
}
