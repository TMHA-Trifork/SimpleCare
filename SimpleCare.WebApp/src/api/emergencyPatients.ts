import { components } from "../types/api-types";
import { getApiUrl } from "./config";

export type EmergencyPatientListItem =
  components["schemas"]["EmergencyPatientListItem"];
export type EmergencyPatient = components["schemas"]["EmergencyPatient"];
export type EmergencyRegistration =
  components["schemas"]["EmergencyRegistration"];
export type TransferRequest = components["schemas"]["TransferRequest"];

export async function fetchEmergencyPatients(
  status?: components["schemas"]["EmergencyPatientStatus"][]
): Promise<EmergencyPatientListItem[]> {
  const url = new URL(getApiUrl("/api/emergency-wards/patients"));
  if (status && status.length > 0) {
    status.forEach((s) => url.searchParams.append("status", s));
  }
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error("Network response was not ok");
  }
  return response.json();
}

export async function fetchEmergencyPatient(
  patientId: string
): Promise<EmergencyPatient> {
  const response = await fetch(
    getApiUrl(`/api/emergency-wards/patients/${patientId}`)
  );
  if (!response.ok) {
    throw new Error("Network response was not ok");
  }
  return response.json();
}

export async function registerEmergencyPatient(
  registration: EmergencyRegistration
): Promise<void> {
  const response = await fetch(
    getApiUrl("/api/emergency-wards/patients/register"),
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(registration),
    }
  );
  if (!response.ok) {
    throw new Error("Network response was not ok");
  }
}

export async function transferEmergencyPatient(
  transfer: TransferRequest
): Promise<void> {
  const response = await fetch(
    getApiUrl("/api/emergency-wards/patients/transfer"),
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(transfer),
    }
  );
  if (!response.ok) {
    throw new Error("Network response was not ok");
  }
}
