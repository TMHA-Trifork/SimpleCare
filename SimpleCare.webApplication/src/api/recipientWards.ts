import { getApiUrl } from "./config";

export type RecipientWard = {
  id: string;
  name: string;
};

export async function fetchRecipientWards(): Promise<RecipientWard[]> {
  const response = await fetch(
    getApiUrl("/api/emergency-wards/recipient-wards")
  );
  if (!response.ok) {
    throw new Error("Network response was not ok");
  }
  return response.json();
}
