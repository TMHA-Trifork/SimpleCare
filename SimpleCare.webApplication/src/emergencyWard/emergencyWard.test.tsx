import { describe, it, expect, vi, Mock } from "vitest";
import { renderWithRouter, fireEvent, act } from "../test/test-utils";
import EmergencyWard from "./emergencyWard";
import * as api from "../api/emergencyPatients";

vi.mock("../api/emergencyPatients");

const mockFetchEmergencyPatients = api.fetchEmergencyPatients as Mock;
const mockRegisterEmergencyPatient = api.registerEmergencyPatient as Mock;

describe("EmergencyWard", () => {
  it("should match snapshot", async () => {
    const mockPatients = [
      {
        patientId: "1",
        givenNames: "John",
        familyName: "Doe",
        status: "Waiting",
      },
    ];

    mockFetchEmergencyPatients.mockResolvedValueOnce(mockPatients);

    let container: HTMLElement;
    await act(async () => {
      const result = renderWithRouter(<EmergencyWard />);
      container = result.container;
    });

    expect(container!.innerHTML).toMatchSnapshot();
  });

  it("should display patient list", async () => {
    const mockPatients = [
      {
        patientId: "1",
        givenNames: "John",
        familyName: "Doe",
        status: "Waiting",
      },
    ];

    mockFetchEmergencyPatients.mockResolvedValueOnce(mockPatients);

    let container: HTMLElement;
    await act(async () => {
      const result = renderWithRouter(<EmergencyWard />);
      container = result.container;
    });

    expect(container!.textContent).toContain("John");
    expect(container!.textContent).toContain("Doe");
  });

  it("should show registration form when toggle button is clicked", async () => {
    mockFetchEmergencyPatients.mockResolvedValueOnce([]);

    let container: HTMLElement;
    await act(async () => {
      const result = renderWithRouter(<EmergencyWard />);
      container = result.container;
    });

    const toggleButton = container!.querySelector(".register-button-toggle");
    expect(toggleButton).toBeTruthy();

    await act(async () => {
      fireEvent.click(toggleButton!);
    });

    const form = container!.querySelector(".register-form");
    expect(form).toBeTruthy();
  });

  it("should register a new patient", async () => {
    mockFetchEmergencyPatients.mockResolvedValueOnce([]);
    mockRegisterEmergencyPatient.mockResolvedValueOnce({});
    mockFetchEmergencyPatients.mockResolvedValueOnce([]);

    const newPatient = {
      givenNames: "Jane",
      familyName: "Smith",
      personalIdentifier: "123456",
      reason: "Emergency",
    };

    let container: HTMLElement;
    await act(async () => {
      const result = renderWithRouter(<EmergencyWard />);
      container = result.container;
    });

    await act(async () => {
      const toggleButton = container!.querySelector(".register-button-toggle");
      fireEvent.click(toggleButton!);
    });

    await act(async () => {
      const inputs = container!.querySelectorAll("input");
      inputs.forEach((input) => {
        fireEvent.change(input, {
          target: {
            name: input.name,
            value: newPatient[input.name as keyof typeof newPatient],
          },
        });
      });
    });

    await act(async () => {
      const form = container!.querySelector(".register-form");
      fireEvent.submit(form!);
    });

    expect(mockRegisterEmergencyPatient).toHaveBeenCalledWith(newPatient);
  });
});
