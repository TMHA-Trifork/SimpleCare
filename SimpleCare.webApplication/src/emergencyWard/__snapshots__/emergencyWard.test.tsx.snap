// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EmergencyWard > should match snapshot 1`] = `"<div class="emergency-ward-container"><h1 class="emergency-ward-title">Emergency Ward</h1><button class="register-button-toggle">Register Patient</button><div class="register-form-container "><form class="register-form"><div class="form-row"><input placeholder="Given Names" required="" type="text" value="" name="givenNames"><input placeholder="Family Name" required="" type="text" value="" name="familyName"></div><div class="form-row"><input placeholder="Personal Identifier" required="" type="text" value="" name="personalIdentifier"><input placeholder="Reason" required="" type="text" value="" name="reason"><button type="submit" class="register-button">Register</button></div></form></div><table class="emergency-ward-table"><thead><tr><th>Given Names</th><th>Family Name</th><th>Status</th><th>Actions</th></tr></thead><tbody><tr class="emergency-ward-row"><td>John</td><td>Doe</td><td>Ukendt</td><td><button class="view-details-button">View Details</button></td></tr></tbody></table></div>"`;
